File: split/multandadd.pas
function multandadd(n: integer; x, y, maxanswer: scaled): scaled;
begin
  if n < 0 then
  begin
    x := -x;
    n := -n;
  end;
  if n = 0 then
    multandadd := y
  else if ((x <= (maxanswer - y) div n) and (-x <= (maxanswer + y) div n)) then
    multandadd := n * x + y
  else
  begin
    aritherror := true;
    multandadd := 0;
  end;
end;

File: cpp/multandadd.cpp
#include "multandadd.h"

scaled multandadd(int n, scaled x, scaled y, scaled maxanswer)
{
	if (n < 0)
	{
		x = -x;
		n = -n;
	}
	if (n == 0)
		return y;
	if (x <= (maxanswer-y)/n && -x <= (maxanswer+y)/n)
		return n*x+y;
	aritherror = true;
	return 0;
}
